contracts/modules/elrond-wasm-module-governance/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+			proposer: proposer.clone(),
+			description,
+			actions: gov_actions,
+		};
+		let proposal_id = self.proposals().push(&proposal);
+
+		let current_block = self.blockchain().get_block_nonce();
+		self.proposal_start_block(proposal_id).set(&current_block);
+
+		self.total_votes(proposal_id).set(&payment_amount);
+		self.votes(proposal_id).insert(proposer, payment_amount);
+
+		Ok(proposal_id)
+	}
+
+	#[payable("*")]
+	#[endpoint]
+	fn vote(
+		&self,
+		#[payment_amount] payment_amount: Self::BigUint,
+		proposal_id: usize,
+	) -> SCResult<()> {
+		self.require_payment_token_governance_token()?;
+		self.require_valid_proposal_id(proposal_id)?;
+		require!(
+			self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Active,
+			"Proposal is not active"
+		);
+
+		let voter = self.blockchain().get_caller();
+
+		self.total_votes(proposal_id)
+			.update(|total_votes| *total_votes += &payment_amount);
+		self.votes(proposal_id)
+			.entry(voter)
+			.and_modify(|nr_votes| *nr_votes += &payment_amount)
+			.or_insert(payment_amount);
+
+		Ok(())
+	}
+
+	#[payable("*")]
+	#[endpoint]
+	fn downvote(
+		&self,
+		#[payment_amount] payment_amount: Self::BigUint,
+		proposal_id: usize,
+	) -> SCResult<()> {
+		self.require_payment_token_governance_token()?;
+		self.require_valid_proposal_id(proposal_id)?;
+		require!(
+			self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Active,
+			"Proposal is not active"
+		);
+
+		let voter = self.blockchain().get_caller();
+
+		self.total_downvotes(proposal_id)
+			.update(|total_downvotes| *total_downvotes += &payment_amount);
+		self.downvotes(proposal_id)
+			.entry(voter)
+			.and_modify(|nr_downvotes| *nr_downvotes += &payment_amount)
+			.or_insert(payment_amount);
+
+		Ok(())
+	}
+
+	#[endpoint]
+	fn queue(&self, proposal_id: usize) -> SCResult<()> {
+		require!(
+			self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Succeeded,
+			"Can only queue succeeded proposals"
+		);
+
+		let current_block = self.blockchain().get_block_nonce();
+		self.proposal_queue_block(proposal_id).set(&current_block);
+
+		Ok(())
+	}
+
+	#[endpoint]
+	fn execute(&self, proposal_id: usize) -> SCResult<()> {
+		require!(
+			self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Queued,
+			"Can only execute queued proposals"
+		);
+
+		let current_block = self.blockchain().get_block_nonce();
+		let lock_blocks = self.lock_time_after_voting_ends_in_blocks().get();
+
+		let lock_start = self.proposal_queue_block(proposal_id).get();
+		let lock_end = lock_start + lock_blocks;
+
+		require!(
+			current_block >= lock_end,
+			"Proposal is in timelock status. Try again later"
+		);
+
+		let proposal = self.proposals().get(proposal_id);
+		let total_gas_needed = self.total_gas_needed(&proposal.actions);
+		let gas_left = self.blockchain().get_gas_left();
+
+		// reserve double to account for the processing needed for each action, like argument serialization
```

Comment:
I'd expect total gas needed to be the total gas needed, not the total gas needed just for action (not including argument serialization). Also remove the comment and try to express the intention with another variable. 
gas_needed_for_action = ...
gas_needed_for_serialization = gas_needed_for_action
total_gas_needed = gas_needed_for_action + gas_needed_for_serialization

Reply:
Fair point. Fixed.

Reply:
Actually, removed the "double" restriction. Will let the user handle the gas costs instead of trying to estimate in the SC. Only restriction is to have more gas than the total the actions need.